src/md.rs

```diff
@@ -1,19 +1,249 @@
+        const ASCII_DIGIT       = 0b0000_0000_0000_0000_0000_1000_0000_0000;
+        const LATIN             = 0b0000_0000_0000_0000_0001_0000_0000_0000;
+        const ALPHABETIC        = 0b0000_0000_0000_0000_0010_0000_0000_0000;
+        const ACCENTUATED       = 0b0000_0000_0000_0000_0100_0000_0000_0000;
+        const CJK               = 0b0000_0000_0000_0000_1000_0000_0000_0000;
+        const HANGUL            = 0b0000_0000_0000_0001_0000_0000_0000_0000;
+        const KATAKANA          = 0b0000_0000_0000_0010_0000_0000_0000_0000;
+        const HIRAGANA          = 0b0000_0000_0000_0100_0000_0000_0000_0000;
+        const THAI              = 0b0000_0000_0000_1000_0000_0000_0000_0000;
+        const CASE_VARIABLE     = 0b0000_0000_0001_0000_0000_0000_0000_0000;
+        const LOWERCASE         = 0b0000_0000_0010_0000_0000_0000_0000_0000;
+        const UPPERCASE         = 0b0000_0000_0100_0000_0000_0000_0000_0000;
+        const NUMERIC           = 0b0000_0000_1000_0000_0000_0000_0000_0000;
+    }
+}
+
+#[derive(Copy, Clone)]
+pub struct MessDetectorChar {
+    pub character: char,
+    pub flags: MessDetectorCharFlags,
+    pub unicode_range: Option<&'static str>,
+}
+
+impl PartialEq for MessDetectorChar {
+    fn eq(&self, other: &Self) -> bool {
+        self.character == other.character
+    }
+}
+
+impl MessDetectorChar {
+    pub fn new(character: char) -> Self {
+        new_mess_detector_character(character)
+    }
+    fn in_category(
+        category: &str,
+        range: Option<&str>,
+        categories_exact: &[&str],
+        categories_partial: &[&str],
+        ranges_partial: &[&str],
+    ) -> bool {
+        // unicode category part
+        if categories_exact.contains(&category)
+            || categories_partial.iter().any(|&cp| category.contains(cp))
+        {
+            return true;
+        }
+        // unicode range part
+        if !ranges_partial.is_empty() {
+            if let Some(range) = range {
+                return ranges_partial.iter().any(|&r| range.contains(r));
+            }
+        }
+        false
+    }
+
+    fn in_description(name: Option<Name>, patterns: &[&str]) -> bool {
+        name.is_some_and(|description| {
+            patterns
+                .iter()
+                .any(|&s| description.to_string().contains(s))
+        })
+    }
+
+    pub fn is(&self, flag: MessDetectorCharFlags) -> bool {
+        self.flags.contains(flag)
+    }
+}
+
+#[cached(
+    type = "UnboundCache<char, MessDetectorChar>",
+    create = "{ UnboundCache::with_capacity(*UTF8_MAXIMAL_ALLOCATION) }",
+    convert = r#"{ character }"#
+)]
+pub fn new_mess_detector_character(character: char) -> MessDetectorChar {
+    let mut flags = MessDetectorCharFlags::empty();
+
+    // PLEASE NOTE! In case of idiomatic refactoring
+    // take in account performance. Sometimes match could be used but it
+    // will require calculate all conditions and can decrease performance
```

Comment:
Also match does not require that all conditions be calculated before hand. 
Both at the rust MIRI level and then LLVM-IR level many optimizations are done.
Very often conditional code are inlined to be faster than if chains, or atleast become equivalent to the if chain in question.

It is true that there are times when if chains are faster than match but that is usually for special cases where inlining and compiler optimizations are impossible
https://github.com/rust-lang/rust/issues/86511

https://rustc-dev-guide.rust-lang.org/mir/optimizations.html
https://rustc-dev-guide.rust-lang.org/backend/lowering-mir.html

Two ways to check is :
- compare the raw assembly of both code
- benchmark.

When this code is integrated, I will try to make this more idiomatic while being guided by the benchmarks